src/Storages/StorageMergeTree.cpp

```diff
@@ -709,6 +713,11 @@ bool StorageMergeTree::merge(
 
         if (!selected)
         {
+            /// If is_single_merged_part is true we treat this part as already merged
+            if (final && is_single_merged_part)
```

Comment:
But `is_single_merged_part` cannot be true unless `final` is also true? Should be `assert(final)` then. In general, returning two bools with only some combinations being valid is a sign of a bad interface. Did you consider making `selectAllPartsToMergeWithinPartition` more explicit about its decisions, e.g. returning an enum that says 'selected some parts / cannot select because of some transient condition (e.g. waiting for other merges) / there is nothing to merge'?

Reply:
This condition would be 'there is nothing to merge', then.